File: Reqnroll/Chapter15/15.2-TheTestDriverPattern/Before/WIMP.Specs/StepDefinitions/OrderingStepDefinitions.cs
using System.Net;
using System.Net.Http.Headers;
using System.Net.Http.Json;

using Reqnroll;

using WIMP.App.Models;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class OrderingStepDefinitions(AuthenticationContext authContext, RestApiContext restApiContext, AppHostingContext appHostingContext)
{
    private int? placedOrderNo;

    [Given("they have placed an order")]
    public async Task GivenTheyHavePlacedAnOrder()
    {
        var placeOrderRequest = new PlaceOrderRequestObjectMother().Build();
        var request = new HttpRequestMessage(HttpMethod.Post, "/api/orders");
        request.Headers.Authorization = new AuthenticationHeaderValue(
            "Bearer", restApiContext.BearerToken);
        request.Content = JsonContent.Create(placeOrderRequest);
        var response = await appHostingContext.AppHost.CreateClient()
            .SendAsync(request);

        Assert.AreEqual(HttpStatusCode.Created, response.StatusCode, $"Place order failed with status code {response.StatusCode}.");

        var placedOrder = await response.Content.ReadFromJsonAsync<Order>()
                    ?? throw new InvalidOperationException("No result payload found");

        placedOrderNo = placedOrder.OrderNo;
    }

    [When("they cancel the placed order")]
    public async Task WhenTheyCancelThePlacedOrder()
    {
        int orderNumber = placedOrderNo ?? throw new InvalidOperationException("No placed order");
        var request = new HttpRequestMessage(HttpMethod.Delete, $"/api/orders/{orderNumber}");
        request.Headers.Authorization = new AuthenticationHeaderValue(
            "Bearer", restApiContext.BearerToken);
        var response = await appHostingContext.AppHost.CreateClient()
            .SendAsync(request);

        Assert.AreEqual(HttpStatusCode.NoContent, response.StatusCode, $"Place order failed with status code {response.StatusCode}.");
    }

    [Then("they should receive a notification about the cancellation")]
    public async Task ThenTheyShouldReceiveANotificationAboutTheCancellation()
    {
        string customerName = authContext.LoggedInCustomerName ?? throw new InvalidOperationException("No logged in customer name");
        var notifications = await appHostingContext.AppHost.CreateClient()
            .GetFromJsonAsync<Notification[]>($"/api/notifications/{customerName}")
            ?? throw new InvalidOperationException("No result payload found");

        Assert.IsNotNull(notifications);
        Assert.IsTrue(notifications.Any(n =>
                n.Message.Contains("cancelled", StringComparison.OrdinalIgnoreCase)),
            "Expected a cancellation notification but none was found.");
    }
}

File: CucumberJVM/Chapter15/15.2-TestDriverPattern/Before/src/test/java/com/wimp/app/specs/stepdefinitions/OrderingStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.models.Order;
import com.wimp.app.specs.support.PlaceOrderRequestObjectMother;
import com.wimp.app.specs.support.RestApiContext;
import io.cucumber.java.en.*;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.client.RestTestClient;
import org.springframework.web.context.WebApplicationContext;

import java.util.Optional;

public class OrderingStepDefinitions {
    private final RestApiContext restApiContext;
    private final RestTestClient restTestClient;

    private Integer placedOrderNo;

    public OrderingStepDefinitions(RestApiContext restApiContext, WebApplicationContext context) {
        this.restApiContext = restApiContext;
        this.restTestClient = RestTestClient.bindToApplicationContext(context).build();
    }

    @Given("they have placed an order")
    public void theyHavePlacedAnOrder() throws Exception {
        var placeOrderRequest = new PlaceOrderRequestObjectMother().build();
        var response = restTestClient.post().uri("/api/orders")
            .accept(MediaType.APPLICATION_JSON)
            .contentType(MediaType.APPLICATION_JSON)
            .header(HttpHeaders.AUTHORIZATION, "Bearer " + restApiContext.getBearerToken())
            .body(placeOrderRequest)
            .exchange();
        response.expectStatus().is2xxSuccessful();
        var placedOrder = Optional.ofNullable(response.expectBody(Order.class).returnResult().getResponseBody())
            .orElseThrow(() -> new RuntimeException("No result payload found"));
        placedOrderNo = placedOrder.getOrderNo();
    }

    @When("they cancel the placed order")
    public void theyCancelTheOrder() throws Exception {
        int orderNo = Optional.ofNullable(placedOrderNo).orElseThrow(() -> new RuntimeException("No placed order"));
        var response = restTestClient.delete().uri("/api/orders/{orderNo}", orderNo)
            .accept(MediaType.APPLICATION_JSON)
            .header(HttpHeaders.AUTHORIZATION, "Bearer " + restApiContext.getBearerToken())
            .exchange();
        response.expectStatus().is2xxSuccessful();
    }
}
